### PR TITLE
Initial Layerwise Profiler

### PR DESCRIPTION
**Pull request by LucasWilkinson**
### SUMMARY:

Initial layerwise profiler leveraging the [kineto](https://github.com/pytorch/kineto) base [PyTorch profiler](https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html). 

NOTE: we run in eager mode by default so that the stack-trace/event-tree contains `nn.Module`, otherwise if cuda-graphs is used all the kernels will be under the `CUDAGraphRunner` module (this is converted to an nn.Module in this PR so that the `_build_module_tree` code puts all those kernels under a `CUDAGraphRunner`)

NOTE: vllm kernels like `vllm::reshape_and_cache_kernel` have no trace or shape informat because they are not registered as a TorchOp (i.e. using `TORCH_LIBRARY`, they instead just a raw `PYBIND11_MODULE` module)

Example on how to use visualization:

```
pip install -r requirements-dev.txt

python examples/offline_profile.py --model nm-testing/OpenHermes-2.5-Mistral-7B-pruned50 --batch-size 4 --prompt-len 512 --json openhermes7b-dense

### For Breakdown Graphs
# For module level breakdown
python neuralmagic/tools/profiler/visualize_trace.py --json-trace openhermes7b-dense.json --output profile_dense.pdf
# For kernel level breakdown
python neuralmagic/tools/profiler/visualize_trace.py --json-trace openhermes7b-dense.json --output profile_dense_kernels.pdf --level kernel

### For table printing
# Decode Summary Table
python neuralmagic/tools/profiler/print_table.py --json-trace openhermes7b-dense.json --phase decode
# Prefill Summary Table
python neuralmagic/tools/profiler/print_table.py --json-trace openhermes7b-dense.json --phase prefill
# Decode Model Layerwise Table
python neuralmagic/tools/profiler/print_table.py --json-trace openhermes7b-dense.json --phase decode --table model
# Prefill Model Layerwise Table
python neuralmagic/tools/profiler/print_table.py --json-trace openhermes7b-dense.json --phase prefill --table model
```

Example Output: [profile-example-output.txt](https://github.com/neuralmagic/nm-vllm/files/14596545/profile-example-output.txt)

### TEST PLAN:
- GHA

### CODE REVIEW

File: examples/offline_profile.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        "filename, will create <filename>_prefill_model_table.csv, "
+        "<filename>_prefill_summary_table.csv, "
+        "<filename>_decode_model_table.csv, and "
+        "<filename>_decode_summary_table.csv")
+    parser.add_argument(
+        "--json",
+        type=str,
+        default=None,
+        help="Export the results as a json file. This should be the filename")
+    parser.add_argument(
```
Comment (robertgshaw2-neuralmagic) [inline examples/offline_profile.py:179]: I would change this to have it such that we call with `--sparsity sparse_w16a16` to match how the server and LLM class are launched

Will eliminate confusion

File: examples/offline_profile.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        "--is_sparse",
+        action='store_true',
+        help=
+        "Use unstructured sparse kernels, sets sparse config to sparse_W16A16")
+    parser.add_argument(
```
Comment (robertgshaw2-neuralmagic) [inline examples/offline_profile.py:184]: I would change this to have it such that we call with `--quantization gptq` to match how the server and LLM class are launched

File: examples/offline_profile.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        type=int,
+        default=PROMPT_LEN_DEFAULT,
+        help=f"Length of the random prompt to use when profiling, all batched "
+        f"requests use the same prompt_len, default={PROMPT_LEN_DEFAULT}")
+    parser.add_argument("--batch_size",
+                        type=int,
+                        default=BATCH_SIZE_DEFAULT,
+                        help=f"Number of requests to run as a single batch, "
+                        f"default={BATCH_SIZE_DEFAULT}")
+    parser.add_argument("--num_gpus",
```
Comment (robertgshaw2-neuralmagic) [inline examples/offline_profile.py:214]: I would update this to `--tensor-parallel-size` to match the args used by vllm

File: requirements-dev.txt [hunk 4]
```diff
@@ -28,3 +28,8 @@ peft
 
 # Benchmarking
 aiohttp
+
+# Profiling
+matplotlib
+pandas
+pyarrow
```
Comment (robertgshaw2-neuralmagic) [inline requirements-dev.txt:35]: nit file

**Review by robertgshaw2-neuralmagic — APPROVED** [review]

File: examples/offline_profile.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+            request_id=f"seq{i}",
+            prompt=None,
+            prompt_token_ids=torch.randint(
+                128,  # 128 to skip over special tokens
+                llm.llm_engine.model_config.get_vocab_size() // 2,
+                size=(prompt_len, )).tolist(),
+            sampling_params=sampling_params)
+
+    with nm_profile() as prefill_prof:
+        llm.llm_engine.step()  # First step is prefill
```
Comment (robertgshaw2-neuralmagic) [inline examples/offline_profile.py:86]: Is this always true?

Does the fact that `max_num_batched_tokens < batch_size * prompt_len` guarantee this?

File: examples/offline_profile.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+        help="The method used to quantize the model weights, "
+        "options are \"marlin\", \"awq\", \"gptq\" and \"squeezellm\"")
+    parser.add_argument(
+        "--max_seq_len",
+        type=int,
+        default=MAX_SEQ_LEN_DEFAULT,
+        help=f"Maximum length of a sequence (including prompt and output), "
+        f"default={MAX_SEQ_LEN_DEFAULT}")
+    parser.add_argument(
+        "--max_num_batched_tokens",
```
Comment (robertgshaw2-neuralmagic) [inline examples/offline_profile.py:197]: Maybe this should be set to `batch_size * prompt_len + eps`?

**Review by robertgshaw2-neuralmagic — APPROVED** [review]

File: examples/offline_profile.py (hunk 3 above)
Comment (LucasWilkinson) [reply examples/offline_profile.py]: Done!

File: examples/offline_profile.py (hunk 2 above)
Comment (LucasWilkinson) [reply examples/offline_profile.py:188]: Done!

File: examples/offline_profile.py (hunk 1 above)
Comment (LucasWilkinson) [reply examples/offline_profile.py:179]: Done!

File: examples/offline_profile.py (hunk 5 above)
Comment (LucasWilkinson) [reply examples/offline_profile.py:86]: as of now yes, but its brittle in the sense that logic could change and this would fall out of sync

**Pull request merged**